nimare/meta/cbma/base.py

```diff
@@ -0,0 +1,480 @@
+
+            # Determine p- and z-values from stat values and null distribution.
+            p_values = np.ones(stat_values.shape)
+            idx = np.where(stat_values > 0)[0]
+            stat_bins = round2(stat_values[idx] * step)
+            p_values[idx] = self.null_distributions_["histogram_weights"][stat_bins]
+
+        elif null_method == "empirical":
+            assert "empirical_null" in self.null_distributions_.keys()
+            p_values = null_to_p(
+                stat_values, self.null_distributions_["empirical_null"], tail="upper"
+            )
+        else:
+            raise ValueError("Argument 'null_method' must be one of: 'analytic', 'empirical'.")
+
+        z_values = p_to_z(p_values, tail="one")
+        return p_values, z_values
+
+    def _p_to_summarystat(self, p, null_method=None):
+        """
+        Compute a summary statistic threshold that corresponds to the provided p-value
+        using either histograms from analytic null or null distribution from empirical
+        null.
+
+        Parameters
+        ----------
+        p : The p-value that corresponds to the summary statistic threshold
+        null_method : {None, "analytic", "empirical"}, optional
+            Whether to use analytic null or empirical null. If None, defaults to using
+            whichever method was set at initialization.
+
+        Returns
+        -------
+        ss : float
+            A float giving the summary statistic value corresponding to the passed p.
+        """
+
+        if null_method is None:
+            null_method = self.null_method
+
+        if null_method.startswith("analytic"):
+            assert "histogram_bins" in self.null_distributions_.keys()
+            assert "histogram_weights" in self.null_distributions_.keys()
+
+            hist_weights = self.null_distributions_["histogram_weights"]
+            # Desired bin is the first one _before_ the target p-value (for consistency
+            # with the empirical null).
+            ss_idx = np.maximum(0, np.where(hist_weights <= p)[0][0] - 1)
+            ss = self.null_distributions_["histogram_bins"][ss_idx]
+
+        elif null_method == "empirical":
+            assert "empirical_null" in self.null_distributions_.keys()
+            null_dist = np.sort(self.null_distributions_["empirical_null"])
+            n_vals = len(null_dist)
+            ss_idx = np.floor(p * n_vals).astype(int)
+            ss = null_dist[-ss_idx]
+        else:
+            raise ValueError("Argument 'null_method' must be one of: 'analytic', 'empirical'.")
+
+        return ss
+
+    def _run_fwe_permutation(self, params):
+        """
+        Run a single Monte Carlo permutation of a dataset. Does the shared work
+        between vFWE and cFWE.
+        """
+        iter_ijk, iter_df, conn, voxel_thresh = params
+
+        iter_ijk = np.squeeze(iter_ijk)
+        iter_df[["i", "j", "k"]] = iter_ijk
```

Comment:
Would it require less memory to copy `self.inputs_["coordinates"]` here instead of copying it in `correct_fwe_montecarlo` and duplicating it `n_iters` times?

Reply:
Hmm... I was going to say that it wouldn't, because there's only one pandas `DataFrame`, and the `n_iters` repetitions are all just references to the same object. But then I realized that when `n_cores > 1`, `imap` *might* end up copying the arguments in each iteration. Though I'm actually not sure—I think it might be copy-on-write, in which case the copying would only happen at the point where we reassign the ijk columns, which is going to happen anyway.

Either way, I think this is likely to be trivial compared to all the other stuff that's going on, so I'm probably not going to muck with it for now. But if we ever do a really serious optimization sweep, it will be worth taking a closer look at.